[CMake] Modernize CMake integration

vicroms opened the pull request:
Hi @ivmai 

I'm a maintainer for [**vcpkg**](https://github.com/Microsoft/vcpkg).

A vcpkg contributor (@NNemec) and I are trying to port `bdwgc` to vcpkg, see: https://github.com/microsoft/vcpkg/pull/6405.

While doing so, I've made some changes to the `CMakeLists.txt` file.

The goal is to match the output of the CMake build to the files distributed in the Debian package (which, I assume is what most downstream consumers use).

Using vcpkg`s CI, I can confirm that this results in successful builds for:

- Windows 32-bit
- Windows 64-bit
- Windows UWP (Windows 10 Store apps)
- Windows for ARM
- x64-Linux (tested with Ubuntu)

The existing `CMakeLists.txt` only produced succesful builds for Windows 32- and 64-bit.

Both the existing version and the modified version fail to build correctly on Mac OS X, but that's something that can be added in a future PR.

**Note: this only produces C++ libraries.**

ivmai commented:
OK, looking into it now. Seems to be good.
The thing I worry about is the style of keywords (upper/lower case)- I want to keep one for all CMake files. I don't have a preference, currently uppercase is used. If you think that lowercase is more readable or more common then it is OK to make a refactory commit.

ivmai commented:
lower-case seems to be more common

ivmai commented:
How about moving cord-specific part to Cmakefile in cord?

NNemec commented:
Sounds reasonable, but it would be good if you could do that as a separate effort to avoid blocking this pull-request. Generally, I would even suggest to deactivate the build of cord by default. My guess is that only a negligible fraction of gc users would be interested in a highly specialized string handling library.

ivmai commented:
> it would be good if you could do that as a separate effort to avoid blocking this pull-request.

OK.

> Generally, I would even suggest to deactivate the build of cord by default. My guess is that only a negligible fraction of gc users would be interested in a highly specialized string handling library.

For now, I'm not going to deactivate it, though I agree that it is of little usage. Currently it is on primarily for gc testing.

ivmai commented:
I will review and merge your patch in 2-3 days

vicroms commented:
I've solved the conflicts with `master` and added back the `add_subdirectory(tests)` which I had removed for the purposes of building the **vcpkg** port.

vicroms commented on CMakeLists.txt:
```diff
@@ -29,7 +29,14 @@ include(CTest)
+option(BDWGC_INSTALL_CORD "Install cord" ON)
+option(BDWGC_SKIP_HEADERS "Skip installing header files" OFF)
+option(BDWGC_SKIP_TESTS "Skip building tests" OFF)
+
```
Any comments on the names of these options?
I namespaced them with a 'BDWGC_' prefix, but these prefix is not present in other options, e.g.: `enable_gc_debug`.

Also, would it be a good idea to move all of the options to the top of the file?

Right now all of the options are declared just before being used, which makes them a bit hard to find.

ivmai replied:
> Any comments on the names of these options?

It is better to remove BDWGC_ prefix.

> Any comments on the names of these options?

Sure. Could you please rebase your PR and do it? (It is OK to do in the same PR or open another one)

vicroms commented on CMakeLists.txt:
```diff
@@ -238,24 +235,59 @@ if (enable_checksums)
-set_target_properties(gcmt-lib PROPERTIES
-                      COMPILE_DEFINITIONS GC_NOT_DLL)
+# Create target
+add_library(gc ${SRC})
```
Usually, in modern CMake, the creation of the target would be near the top of the file.

That way, definitions can be added with `target_add_definitions()` to specific targets instead of `add_definitions()` which is global.

ivmai replied:
OK, please add a commit or PR.

vicroms replied:
Will do in a separate PR after this one is finished

vicroms commented on CMakeLists.txt:
```diff
@@ -238,24 +235,59 @@ if (enable_checksums)
+# Install library
+install(
+    TARGETS gc
+    LIBRARY DESTINATION lib 
```
A feature to consider would be exporting the target, so that this CMakeLists generates `gc-config.cmake` and `gc-targets.cmake`, which would make linking to `gc` easier for downstream consumers. i.e.: it would allow `find_package(gc)` to work out-of-the-box.

ivmai replied:
OK

vicroms commented on CMakeLists.txt:
```diff
@@ -238,24 +235,59 @@ if (enable_checksums)
 endif()
 
-add_subdirectory(tests)
+if (NOT BDWGC_SKIP_TESTS)
```
I added an option to control whether tests are installed here.
It is ON by default.

ivmai replied:
Having an option is OK.
Test generation should be on by default. Is it OK?

vicroms commented on CMakeLists.txt:
```diff
@@ -29,7 +29,14 @@ include(CTest)
 
 cmake_minimum_required(VERSION 2.6)
```
Would you consider raising the minimum required version to 3.0?

There's much added goodness that we're missing on like `CMakePackageConfigHelpers` for easy integration in CMake projects.

ivmai replied:
OK, please make a commit or PR

ivmai commented:
I will reply to your questions later.

What's wrong with https://ci.appveyor.com/project/ivmai/bdwgc/build/job/j1450qvf94bqjary ?

vicroms commented:
> What's wrong with https://ci.appveyor.com/project/ivmai/bdwgc/build/job/j1450qvf94bqjary ?

I'm not too familiar with how testing works, but it looks like the tests are not linking `gc` properly because they reference the old target name `gc-lib`.

I'll update the `tests/CMakeLists.txt` to use the new target name.

ivmai commented:
I'm continuing review of your 2 patches in this PR.

ivmai commented on CMakeLists.txt:
```diff
@@ -88,10 +92,10 @@ if (CMAKE_USE_PTHREADS_INIT)
   # Common defines for most POSIX platforms.
   if (HOST MATCHES .*-.*-aix.*|.*-.*-android.*|.*-.*-cygwin.*|.*-.*-darwin.*|.*-.*-.*freebsd.*|.*-.*-haiku.*|.*-.*-gnu.*|.*-.*-hpux11.*|.*-.*-irix.*|.*-.*-.*linux.*|.*-.*-msys.*|.*-.*-nacl.*|.*-.*-netbsd.*|.*-.*-openbsd.*|.*-.*-osf.*|.*-.*-solaris.*)
     add_definitions("-DGC_THREADS -D_REENTRANT")
-    if (enable_parallel_mark)
```
I'm going to leave lower case variable names for now (to match the configure)

vicroms replied:
Got it.
Merged master into PR branch and used all lowercase names for options

ivmai commented:
This pull request **introduces 11 alerts** when merging af50acc98de6e7690ead61b80b7601591739d4c0 into 2b7003769e1ecc941ef80b603cba16a121f37997 - [view on LGTM.com](https://lgtm.com/projects/g/ivmai/bdwgc/rev/pr-97377805945f82c495a614f6684a4a5a808f4bc0)

**new alerts:**

* 10 for FIXME comment
* 1 for Missing header guard

---

*Comment posted by [LGTM.com](https://lgtm.com)*

vicroms commented:
I noticed that one of the CI tests fails with the following error:

```
libtool: link: gcc -DSTATICROOTSLIB2 -DGC_VISIBILITY_HIDDEN_SET -fvisibility=hidden -Wall -Wextra -Wpedantic -Wno-long-long -g -O2 -fno-strict-aliasing -D GCTEST_PRINT_VERBOSE -o .libs/staticrootstest.exe tests/staticrootstest-staticrootstest.o  ./.libs/libstaticrootslib_test.dll.a ./.libs/libstaticrootslib2_test.dll.a /cygdrive/c/projects/bdwgc/.libs/libgc.dll.a ./.libs/libgc.dll.a -L/nowhere -L/usr/local/lib
```

If you notice, it is liking to the old binary names `libgcc.dll.a`, while this CMakeLists should produce:
- `gc.dll`, and 
- `gc.lib` 
on Windows.

ivmai commented:
What does it mean "install tests"? Why do we need an option for it?

vicroms commented:
It is ON by default, if it is set to OFF then it skips building the tests.  It is useful in **vcpkg** so that only the library is installed without any other extra binary (like cord for example).

ivmai commented:
Review is still in progress.

Is it OK if I rename install_test, install_cord to build_test/cord ?

vicroms commented:
Yes, that's ok.

ivmai commented on CMakeLists.txt:
```diff
@@ -52,22 +56,13 @@ add_definitions("-DALL_INTERIOR_POINTERS -DNO_EXECUTE_PERMISSION")
 if (APPLE)
   if ("${CMAKE_OSX_ARCHITECTURES}" STREQUAL "")
-    set(CMAKE_OSX_ARCHITECTURES "ppc;i386;x86_64"
+    set(CMAKE_OSX_ARCHITECTURES "x86_64;i386"
```
Does the presence of ppc lead to build failure on OS X?

vicroms replied:
I'm not a Mac user myself, but this change enables Mac OS on **vcpkg**'s CI tests.

The following snippet causes a build failure when adding the `ppc` architecture:  

[`cdefs.h`](https://opensource.apple.com/source/xnu/xnu-4570.41.2/bsd/sys/cdefs.h.auto.html)
```
/*
 * Architecture validation for current SDK
 */
#if !defined(__sys_cdefs_arch_unknown__) && defined(__i386__)
#elif !defined(__sys_cdefs_arch_unknown__) && defined(__x86_64__)
#elif !defined(__sys_cdefs_arch_unknown__) && defined(__arm__)
#elif !defined(__sys_cdefs_arch_unknown__) && defined(__arm64__)
#else
#error Unsupported architecture
#endif
```

I assume that this only fails newer Mac OS versions.
I'm not sure in which version of MacOS was `ppc` support deprecated, but it is worth investigating so that we can conditionally pass the `ppc` architecture.

It is based on @nickwanniger 's fix for issue #275

ivmai replied:
OK. It would be later to add arm and arm64 (when someone tests them)

ivmai replied:
OK. It would be later to add arm and arm64 (when someone tests them).

nickwanninger commented:
I mentioned in #275, that if you run `cmake .. -DCMAKE_OSX_ARCHITECTURES=x86_64`, it seems to work with macOS, ignoring a warning:
```
CMake Warning (dev):
  Policy CMP0042 is not set: MACOSX_RPATH is enabled by default.  Run "cmake
  --help-policy CMP0042" for policy details.  Use the cmake_policy command to
  set the policy and suppress this warning.

  MACOSX_RPATH is not specified for the following targets:

   gcmt-dll

This warning is for project developers.  Use -Wno-dev to suppress it.
```

But other than that, this PR works on macOS with that flag

ivmai commented:
I have an issue with tests execution when shared library is built (at least on Windows): the tests are created and executed in tests folder while the library is created in base folder, so the tests fail because the library is not found.
I see 3 ways to fix it:
1. move the content of tests/CMakeLists.txt to base CMakeLists.txt
2. tell cmake somehow that the test executables should be put to the base folder
3. tell cmake/ctest somehow (working_directory?) that the tests should be launched from the base folder

Ideal way is 2 but I failed to do it; neither way 3. It could be fixed with way 1 but I'd like to keep tests/CMakeLists.txt.

Do you know how to fix it in way 2 or 3?

wangp commented:
Can you use `add_test(... WORKING_DIRECTORY ...)` or set the PATH environment variable?
https://cmake.org/pipermail/cmake/2015-November/061978.html
(not tested, sorry)

vicroms commented:
Sorry, I haven't had time to look into the issues.
Any updates so far?

ivmai commented:
In progress. I'll finish this week

ivmai commented:
Why do we need target_include_directories ?

I have pushed all the rest of the changes your requested. Please check it.

ivmai commented:
Related PR: https://github.com/microsoft/vcpkg/pull/20580 (update package to v8.2.0)
